custom/enikshay/nikshay_datamigration/management/commands/create_enikshay_cases.py

```diff
@@ -0,0 +1,283 @@
+                'case_type': 'test',
+                'update': {
+                    'date_tested': followup.TestDate,
+
+                    'migration_created_case': True,
+                    'migration_followup_id': followup.id,
+                },
+            },
+            'indices': [CaseIndex(
+                occurrence_structure,
+                identifier='host',
+                relationship=CASE_INDEX_EXTENSION,
+                related_type=occurrence_structure.attrs['case_type'],
+            )],
+        }
+
+        for test_case in self.case_accessor.get_cases([
+            index.referenced_id for index in
+            self.case_accessor.get_case(occurrence_structure.case_id).reverse_indices
+        ]):
+            dynamic_case_properties = test_case.dynamic_case_properties()
+            if (
+                'migration_followup_id' in dynamic_case_properties
+                and followup.id == int(test_case.dynamic_case_properties()['migration_followup_id'])
+            ):
+                kwargs['case_id'] = test_case.case_id
+                kwargs['attrs']['create'] = False
+
+        return CaseStructure(**kwargs)
+
+    @property
+    @memoized
+    def _outcome(self):
+        zero_or_one_outcomes = list(Outcome.objects.filter(PatientId=self.patient_detail))
+        if zero_or_one_outcomes:
+            return zero_or_one_outcomes[0]
+        else:
+            return None
+
+    @property
+    @memoized
+    def _followups(self):
+        return Followup.objects.filter(PatientID=self.patient_detail)
+
+    @property
+    def _location(self):
+        return self.nikshay_code_to_location(self.domain)[self._nikshay_code]
+
+    @classmethod
+    @memoized
+    def nikshay_code_to_location(cls, domain):
+        return {
+            location.metadata.get('nikshay_code'): location
+            for location in SQLLocation.objects.filter(domain=domain)
+            if 'nikshay_code' in location.metadata
+        }
+
+    @property
+    def _nikshay_code(self):
+        return '-'.join(self.patient_detail.PregId.split('-')[:4])
+
+
+class Command(BaseCommand):
+
+    def handle(self, domain, **options):
+        base_query = PatientDetail.objects.all()
+
+        start = options['start']
+        limit = options['limit']
+
+        if limit is not None:
+            patient_details = base_query[start:start + limit]
+        else:
+            patient_details = base_query[start:]
+
+        total = patient_details.count()
+        counter = 0
+        num_succeeded = 0
+        num_failed = 0
+        logger.info('Starting migration of %d patient cases.' % total)
+        for patient_detail in patient_details:
+            counter += 1
+            try:
+                case_factory = EnikshayCaseFactory(domain, patient_detail)
+                case_factory.create_cases()
+            except:
+                num_failed += 1
+                logger.error('Failed on %d of %d.' % (counter, total), exc_info=True)
```

Comment:
one thing I sometimes think about is how these log files can be parsed/worked with in an automated way. not sure exactly how you plan to consume this, but if it's anything other than reading the log file it might make sense to try and structure a little more rigidly